homeassistant/components/todo/__init__.py

```diff
@@ -0,0 +1,310 @@
+
+
+WEBSOCKET_TODO_ITEM_CREATE_SCHEMA = vol.Schema(
+    {
+        vol.Required("summary"): cv.string,
+        vol.Optional("status", default=TodoItemStatus.NEEDS_ACTION): vol.In(
+            {TodoItemStatus.NEEDS_ACTION, TodoItemStatus.COMPLETED}
+        ),
+    },
+)
+WEBSOCKET_TODO_ITEM_UPDATE_SCHEMA = vol.Schema(
+    {
+        vol.Required("uid"): cv.string,
+        vol.Required("summary"): cv.string,
+        vol.Optional("status"): vol.In(
+            {TodoItemStatus.NEEDS_ACTION, TodoItemStatus.COMPLETED}
+        ),
+    },
+)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Todo entities."""
+    component = hass.data[DOMAIN] = EntityComponent[TodoListEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_list)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_create)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_delete)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_update)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_move)
+
+    await component.async_setup(config)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclasses.dataclass
+class TodoItem:
+    """A To-do item in a To-do list."""
+
+    summary: str
+    """The summary that represents the item."""
+
+    uid: str | None = None
+    """A unique identifier for the To-do item."""
+
+    status: TodoItemStatus = TodoItemStatus.NEEDS_ACTION
+    """A status or confirmation of the To-do item."""
+
+
+class TodoListEntity(Entity):
+    """An entity that represents a To-do list."""
+
+    @property
+    def state(self) -> int | None:
+        """Return the entity state as the count of incomplete items."""
+        items = self.todo_items
+        if items is None:
+            return None
+        return sum([item.status == TodoItemStatus.NEEDS_ACTION for item in items])
+
+    @property
+    @abstractmethod
+    def todo_items(self) -> list[TodoItem] | None:
+        """Return the To-do items in the To-do list."""
+
+    async def async_create_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_update_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_delete_todo_items(self, uids: set[str]) -> None:
```

Comment:
Do we use a set parameter with multiple uids to follow the rfc spec or why do we use multiple instead of a single item for this action?

Reply:
This is purely for practical performance concerns, not related to the rfc spec.

When reviewing the shopping list API, one observation was that we currently require a "clear completed items" for the shopping list.  This is also part of the new spec for todo when I looked at the wireframes shared by Bram that Madelena made.  (This is mentioned in https://github.com/home-assistant/architecture/discussions/962 in passing under "Consequences" as a performance topic)

An alternative could be to rename this "clear completed items" rather than supporting delete of arbitrary items, however then i believe it requires more custom logic by the integration to implement. I liked simplifying this by pushing this to be a frontend concern about which items it is deleting (delete is needed anyway) allowing cloud integrations to have a more straight forward API. 

Open to suggestions.